sopel/irc/modes.py

```diff
@@ -0,0 +1,234 @@
+"""Mode management for IRC channels.
+
+.. seealso::
+
+    https://modern.ircdocs.horse/#mode-message
+
+.. versionadded:: 8.0
+"""
+from __future__ import generator_stop
+
+from collections import namedtuple
+import logging
+from typing import Dict, Generator, List, Optional, Set, Tuple
+
+
+LOGGER = logging.getLogger(__name__)
+
+PARAM_ALWAYS = 'always'
+PARAM_ADDED = 'added'
+PARAM_REMOVED = 'removed'
+PARAM_NEVER = 'never'
+DEFAULT_MODETYPE_PARAM_CONFIG = {
+    'A': PARAM_ALWAYS,
+    'B': PARAM_ALWAYS,
+    'C': PARAM_ADDED,
+    'D': PARAM_NEVER,
+}
+"""Default parameter requirements for mode types."""
+
+
+class ModeException(Exception):
+    """Base exception class for mode management."""
+    pass
+
+
+class ModeTypeUnknown(ModeException):
+    """Exception when a mode's type is unknown or cannot be determined."""
+    def __init__(self, mode) -> None:
+        super().__init__('Unknown type for mode %s' % mode)
+
+
+class ModeTypeImproperlyConfigured(ModeException):
+    """Exception when the mode's type management is not configured."""
+    def __init__(self, mode: str, letter: str) -> None:
+        message = 'Type {mode} for mode {letter} is not properly configured.'
+        super().__init__(message.format(mode=mode, letter=letter))
+
+
+def _modes_is_added(modestring: str
+                    ) -> Generator[Tuple[str, bool], None, None]:
+    is_added = True
+    for char in modestring:
+        if char in '+-':
+            is_added = char == '+'
+            continue
+        yield (char, is_added)
+
+
+ModeString = namedtuple('ModeString', [
+    'modes',
+    'privileges',
+    'ignored_modes',
+    'leftover_params',
+])
+
+
+class ModeMessage:
+    """Modestring parser for IRC's ``MODE`` messages for channel modes."""
```

Comment:
Actually... yes it could. I just read that user modes are of type D (no param), so that should be doable.

Reply:
14 days later, and I think everyone saw that one coming: that will be (if ever) for another PR.

After all, parsing user mode is much easier by being all of type D, and since Sopel doesn't do anything with them, for now, I'll keep this PR focused on channel modes only.